Add examples to `key` module

tcharding opened the pull request:
This PR is an initial attempt to more thoroughly test our public API.

Add examples to various types/methods/functions in the key module.

I'm not entirely sure when is enough, do we want an example on every single public method, function, and type or is this overkill. In this PR I tried to find a balance by doing ever method/function that took an argument that is a custom type from this lib. I think this should be extended to include return values too though ...

Thanks to @thomaseizinger for the idea!

First 2 patches are docs improvements to `lib.rs`.

tcharding commented:
oh poo - was running tests without `--features=rand`, converting to draft so I can fix this up tomorrow.

elichai commented on src/key.rs:
```diff
@@ -26,7 +27,22 @@ use Verification;
+/// use secp256k1::rand::rngs::OsRng;
+///
+/// let secp = Secp256k1::new();
+/// let mut rng = OsRng::new().expect("OsRng");
```
Hmm why not `thread_rng()`?

tcharding replied:
Because we use `thread_rng` everywhere else, I thought it might be nice to use OsRng here, although its already used in `lib.rs` example so doesn't really add that much.

elichai commented on src/key.rs:
````diff
@@ -113,7 +150,16 @@ impl SecretKey {
+    /// # Examples
+    ///
+    /// ```
+    /// # #[cfg(feature="rand")] {
+    /// use secp256k1::{rand, SecretKey};
````
rand isn't really used here

tcharding replied:
Oh cheers, I've not worked out how to lint the examples yet.

thomaseizinger replied:
I haven't tried it but I'd expect `--all-targets` on `cargo clippy` to include rustdoc.

tcharding replied:
I only learned during this PR that `-all-targets` does _not_ include docs (I have my editor configured to run clippy with `--all-targets`, thanks Lucas :)

elichai commented on src/key.rs:
````diff
@@ -135,7 +181,20 @@ impl SecretKey {
+    ///
+    /// ```
+    /// # #[cfg(feature="rand")] {
+    /// use secp256k1::{rand, Secp256k1, SecretKey, KeyPair};
````
the `rand` import here is unnecessary

tcharding replied:
Cheers

elichai commented on src/key.rs:
````diff
@@ -411,17 +503,53 @@ impl PublicKey {
+    /// ```
+    /// # #[cfg(feature="rand")] {
+    /// use secp256k1::Secp256k1;
+    /// use secp256k1::rand::{RngCore, thread_rng};
````
Is `RngCore` really needed here?

tcharding replied:
Nope, thanks.

elichai commented on src/key.rs:
```diff
@@ -411,17 +503,53 @@ impl PublicKey {
+    /// let mut rng = thread_rng();
+    /// let (_, pk1) = secp.generate_keypair(&mut rng);
+    /// let (_, pk2) = secp.generate_keypair(&mut rng);
+    /// let sum = pk1.combine(&pk2).expect("sum");
```
Maybe replace `.expect("sum")` with `unwrap` or `expect("It's improbable to fail for 2 random public keys")`

elichai commented on src/key.rs:
```diff
@@ -511,27 +639,42 @@ impl Ord for PublicKey {
+/// use secp256k1::rand::rngs::OsRng;
+///
+/// let secp = Secp256k1::new();
+/// let mut rng = OsRng::new().expect("OsRng");
```
Why not `thread_rng`?

tcharding replied:
I'm going to remove all the `OsRng`s and use `thread_rng` througout. Will do as a separate patch.

elichai commented on src/key.rs:
```diff
@@ -612,20 +766,38 @@ impl KeyPair {
+    /// thread_rng().fill_bytes(&mut tweak);
+    ///
+    /// let mut key_pair = KeyPair::new(&secp, &mut thread_rng());
+    /// key_pair.tweak_add_assign(&secp, &tweak).expect("Tweak error");
```
I don't think `"Tweak error"` is any more helpful than just an unwrap here.
so I think it should either be `unwrap` or  `"Improbable to fail with a randomly generated tweak"`

tcharding replied:
Used the descriptive string suggested, thanks.

elichai commented on src/key.rs:
```diff
@@ -837,15 +1024,34 @@ impl XOnlyPublicKey {
+    ///
+    /// let mut key_pair = KeyPair::new(&secp, &mut thread_rng());
+    /// let mut public_key = key_pair.public_key();
+    /// public_key.tweak_add_assign(&secp, &tweak).expect("Tweak error");
```
Same comment about the expect

elichai commented on src/key.rs:
```diff
@@ -882,18 +1088,37 @@ impl XOnlyPublicKey {
+    /// let mut public_key = key_pair.public_key();
+    /// let original = public_key;
+    /// let parity = public_key.tweak_add_assign(&secp, &tweak).expect("Tweak error");
+    /// assert!(original.tweak_add_check(&secp, &public_key, parity, tweak));
```
Too bad we don't have access to a hash function to showcase this properly :(

elichai commented:
Sorry for the long review, it's easy to bikeshed over documentation

tcharding commented:
> Sorry for the long review, it's easy to bikeshed over documentation

No apology needed, I appreciate the comments. Documentation is definitely a funny thing, its easy to agree we need it not so easy to agree how it should be done. That's why I tried to stick meticulously to Rust conventions. (I've not written that much thorough Rust documentation in the past though so I'm still learning.)

tcharding commented:
@elichai I fixed all the 'should have been found by the linter' issues and squashed them into  the original commit (`4fc759a Add examples to types and methods in key module`). The `OsRng` suggestion is done as a separate patch along with changing how `rand::thread_rng` is called, please say if you disagree with that patch. Thanks.

thomaseizinger commented on src/key.rs:
````diff
@@ -26,7 +27,20 @@ use Verification;
+/// Basic usage:
+///
+/// ```
+/// # #[cfg(feature="rand")] {
+/// use secp256k1::{rand, Secp256k1, SecretKey};
+///
+/// let secp = Secp256k1::new();
+/// let secret_key = SecretKey::new(&mut rand::thread_rng());
+/// # }
````
Nice idea with the scope. Untested idea that might also work (note the `#!`):

```suggestion
/// # #![cfg(feature="rand")]
/// use secp256k1::{rand, Secp256k1, SecretKey};
///
/// let secp = Secp256k1::new();
/// let secret_key = SecretKey::new(&mut rand::thread_rng());
```

tcharding replied:
This doesn't work with Rust 1.29 unfortunately. I think, IIRC, it starts working at 1.34. I'll add a TODO to remove them after bump of MSRV.

tcharding replied:
Fail! I did not 'note the `!#`' even when you told me too. It is _hard_ to get good help :)
Updating this PR to include this suggestion.

tcharding replied:
Bother, and I thought the saying was 'if it builds ship it' ... turns out the `!#` suggestion builds just fine but doesn't run.

thomaseizinger replied:
Oh that is weird. Should go back to testing my suggestions before making them!

tcharding replied:
Not at all, open suggestions lead to leaning stuff. And since they are cheaper to make than tested suggestions it leads to more information being shared, IMO.

thomaseizinger commented on src/key.rs:
```diff
@@ -411,17 +496,51 @@ impl PublicKey {
-    /// an error if the result would be the point at infinity, i.e. we are adding
-    /// a point to its own negation, if the provided slice has no element in it,
-    /// or if the number of element it contains is greater than i32::MAX.
+    /// Adds the keys in the provided slice together, returning the sum.
```
Perhaps this example should use three keys instead of two?

tcharding replied:
It crossed my mind during dev, if you thought it too lets do it. Added as a separate patch.

thomaseizinger reviewed (commented):
Nice work.

tcharding commented:
After re-visiting this I've squashed all the suggested changes into the one commit `e9699a1 Add examples to types and methods in key module`.

apoelstra reviewed (approved):
ACK aa828f01a5727f0b056975cdbcaedbb24efb78bd

This is amazing, thanks!

tcharding commented:
Cheers man!
